coredump: parse build-id and .note.package

bluca opened the pull request:
Initial implementation for #18433 - the binary is stamped using the linker script from https://github.com/keszybz/rpm-version-note/blob/main/generate-package-notes.py via `LDFLAGS="-Wl,-T,$PWD/script"`.

The build-id for each ELF module in a core file is also parsed.

This is still incomplete, as parsing .note.package requires the actual executable being available and readable by systemd-coredump. The note is part of the core file, but currently there not seem to be a known way to parse that via elfutils' libraries. I am working on this with the elfutils maintainers (Mark Wielaard and Frank Eigler, a huge thank you to them for helping to figure this out).
EDIT: this is now working, thanks again to Mark and Frank for the help!

I am opening this as we can start to think about what to do with this metadata. At the moment it's just printed in the journal as such:

```
Process 738465 (fsverity) of user 1000 dumped core.

Found module linux-vdso.so.1 with build-id: 5d90781f8a5a7cfa89af4cfe67362f166a969137
Found module ld-linux-x86-64.so.2 with build-id: f25dfd7b95be4ba386fd71080accae8c0732b711
Found module libpthread.so.0 with build-id: e91114987a0147bd050addbd591eb8994b29f4b3
Found module libdl.so.2 with build-id: d3583c742dd47aaa860c5ae0c0c5bdbcd2d54f61
Found module libc.so.6 with build-id: 18b9a9a8c523e5cfe5b5d946d605d09242f09798
Found module libcrypto.so.1.1 with build-id: 749142d5ee728a76e7cdc61fd79d2311a77405a2
Found module fsverity with build-id: a37208c69a0a9e07979632fed09d23140ddb0276
Metadata for module fsverity owned by TBD found: {"packageType":"deb","package":"fsverity-utils","packageVersion":"1.3-1"}
Stack trace of thread 738465:
#0  0x00007fd82a01d6f4 __GI___nanosleep (libc.so.6 + 0xc66f4)
#1  0x00007fd82a01d62a __sleep (libc.so.6 + 0xc662a)
#2  0x00005558a248b07d main (fsverity + 0x307d)
#3  0x00007fd829f7b09b __libc_start_main (libc.so.6 + 0x2409b)
#4  0x00005558a248b1ea _start (fsverity + 0x31ea)
```

keszybz reviewed (commented):
Looks are reasonable...

bluca commented on src/coredump/stacktrace.c:
```diff
@@ -150,6 +151,62 @@ static int module_callback (Dwfl_Module *mod, void **userdata, const char *name,
+                         * magic ID is always the same. */
+                        if (note_header.n_type == 0x7add3300)
+                                fprintf(c->f, "Metadata for module %s owned by %s found: %s\n",
+                                        name, note_name, payload);
```
Yes - this is just temporary, as I wasn't sure we want to add it like this to the journal or not. If we want to, then I'll pretty-print it

bluca replied:
done, and added to journal metadata too

poettering commented on src/coredump/stacktrace.c:
```diff
@@ -111,10 +113,51 @@ static int thread_callback(Dwfl_Thread *thread, void *userdata) {
+         * stores it separately in an internal library struct. */
+        id_len = dwfl_module_build_id(mod, &id, &id_vaddr);
+        if (id_len == 0) {
+                fprintf(c->f, " without build-id\n");
```
fputs()?

poettering commented on src/coredump/stacktrace.c:
```diff
@@ -111,10 +113,51 @@ static int thread_callback(Dwfl_Thread *thread, void *userdata) {
+
+        id_hex = hexmem(id, id_len);
+        if (!id_hex) {
+                fprintf(c->f, "\n");
```
fputc()

poettering commented on src/coredump/stacktrace.c:
```diff
@@ -150,6 +151,62 @@ static int module_callback (Dwfl_Module *mod, void **userdata, const char *name,
+                Elf_Data *data;
+
+                /* Package metadata is in PT_NOTE headers */
+                program_header = gelf_getphdr (elf, i, &mem);
```
spurious space between function name and ()

poettering commented on src/coredump/stacktrace.c:
```diff
@@ -150,6 +151,62 @@ static int module_callback (Dwfl_Module *mod, void **userdata, const char *name,
+
+                /* Package metadata is in PT_NOTE headers */
+                program_header = gelf_getphdr (elf, i, &mem);
+                if (program_header == NULL || program_header->p_type != PT_NOTE)
```
!program_header

poettering commented on src/coredump/stacktrace.c:
```diff
@@ -150,6 +151,62 @@ static int module_callback (Dwfl_Module *mod, void **userdata, const char *name,
+                                continue;
+                        /* Package metadata might have different owners, but the
+                         * magic ID is always the same. */
+                        if (note_header.n_type == 0x7add3300)
```
maybe add a `#define` for this

bluca replied:
done

poettering commented:
hmm, the package json data probably needs somedistro metadata, too.

i.e. we want to have enough metadata in there so that we can clearly pinpoint which RPM/DEB a crash is from. i.e. which entry from https://www.rpmfind.net/linux/rpm2html/search.php?query=systemd is the relevant one, in case systemd crashes.

this means we might some cpu ID too (i.e. the ELF arch might differ from the RPM cpu if they pick a higher baseline of the same arch)

bluca commented:
> hmm, the package json data probably needs somedistro metadata, too.
> 
> i.e. we want to have enough metadata in there so that we can clearly pinpoint which RPM/DEB a crash is from. i.e. which entry from https://www.rpmfind.net/linux/rpm2html/search.php?query=systemd is the relevant one, in case systemd crashes.
> 
> this means we might some cpu ID too (i.e. the ELF arch might differ from the RPM cpu if they pick a higher baseline of the same arch)

Sounds good to me - could you please mention that on https://github.com/systemd/systemd/issues/18433

bluca commented:
I am happy now with how this looks - we can parse the metadata even if we don't have access to the executable.

The following metadata is added to the coredump journal log:

```
    COREDUMP_PKGMETA_FSVERITY_PACKAGETYPE=deb
    COREDUMP_PKGMETA_FSVERITY_PACKAGE=fsverity-utils
    COREDUMP_PKGMETA_FSVERITY_PACKAGEVERSION=1.3-1
```

The log message looks like:

```
    MESSAGE=Process 1185643 (fsverity) of user 1000 dumped core.
            
            Found module linux-vdso.so.1 with build-id: 5d90781f8a5a7cfa89af4cfe67362f166a969137
            Found module ld-linux-x86-64.so.2 with build-id: f25dfd7b95be4ba386fd71080accae8c0732b711
            Found module libpthread.so.0 with build-id: e91114987a0147bd050addbd591eb8994b29f4b3
            Found module libdl.so.2 with build-id: d3583c742dd47aaa860c5ae0c0c5bdbcd2d54f61
            Found module libc.so.6 with build-id: 18b9a9a8c523e5cfe5b5d946d605d09242f09798
            Found module libcrypto.so.1.1 with build-id: 749142d5ee728a76e7cdc61fd79d2311a77405a2
            Found module fsverity with build-id: d5d21eac3b97c2562a65fd87973f8e1e09323995
            Metadata for module fsverity owned by TBD found: {
            	"packageType" : "deb",
            	"package" : "fsverity-utils",
            	"packageVersion" : "1.3-1"
            }
            
            Stack trace of thread 1185643:
            #0  0x00007fcbcc9a56f4 __GI___nanosleep (libc.so.6 + 0xc66f4)
            #1  0x00007fcbcc9a562a __sleep (libc.so.6 + 0xc662a)
            #2  0x0000559cd6ab007d main (fsverity + 0x307d)
            #3  0x00007fcbcc90309b __libc_start_main (libc.so.6 + 0x2409b)
            #4  0x0000559cd6ab01ea _start (fsverity + 0x31ea)
```

bluca commented:
Open questions:

- is this how we want to log metadata, with COREDUMP_PKGMETA_<module>_<json key>=<json value> or is there a better way?
- how do I parse COREDUMP_PKG_META_* (wildcard) in coredumpctl?

bluca commented:
Added the build-id to the journal metadata as well.

keszybz commented:
Hmm, `COREDUMP_PKGMETA_FSVERITY_PACKAGETYPE=...`. The "FSVERITY" part comes from "module fsverity". Is the plan to allow multiple such entry lists (`_PACKAGETYPE` + `_PACKAGE` + `_PACKAGE_VERSION`)? If feels strange to include this information in the field *name*, instead of the field *value*. Also, there's the problem that the character set is very limited. So what happens when I have "module foobar" linked with "module foobar++" (e.g. C and C++ parts). We can't express them together. Also, how does one figure out what field to look up knowing the "module name"?

A related question: I assumed we'd be able tell the user the "main" package name, i.e. package of the crashing executable. In this scheme, if there are multiple modules with metadata, can the consumer say which is the "main" one?

I'd expect something like
```
COREDUMP_PKGMETA_PACKAGETYPE=xxx:deb
COREDUMP_PKGMETA_PACKAGE=xxx:fsverity-utils
COREDUMP_PKGMETA_PACKAGEVERSION=xxx:1.3-1
COREDUMP_PKGMETA_BUILDID=xxx:d5d21eac3b97c2562a65fd87973f8e1e09323995
```
where `xxx` is either the "module name" or just a number, e.g. "1" for the first module, and then "2", "3", "4"… for the libraries. Numbers have the advantage that we could agree that "1" is always the executable.

bluca commented:
> Hmm, `COREDUMP_PKGMETA_FSVERITY_PACKAGETYPE=...`. The "FSVERITY" part comes from "module fsverity". Is the plan to allow multiple such entry lists (`_PACKAGETYPE` + `_PACKAGE` + `_PACKAGE_VERSION`)? If feels strange to include this information in the field _name_, instead of the field _value_. Also, there's the problem that the character set is very limited. So what happens when I have "module foobar" linked with "module foobar++" (e.g. C and C++ parts). We can't express them together. Also, how does one figure out what field to look up knowing the "module name"?
>
> A related question: I assumed we'd be able tell the user the "main" package name, i.e. package of the crashing executable. In this scheme, if there are multiple modules with metadata, can the consumer say which is the "main" one?
> 
> I'd expect something like
> 
> ```
> COREDUMP_PKGMETA_PACKAGETYPE=xxx:deb
> COREDUMP_PKGMETA_PACKAGE=xxx:fsverity-utils
> COREDUMP_PKGMETA_PACKAGEVERSION=xxx:1.3-1
> COREDUMP_PKGMETA_BUILDID=xxx:d5d21eac3b97c2562a65fd87973f8e1e09323995
> ```
> 
> where `xxx` is either the "module name" or just a number, e.g. "1" for the first module, and then "2", "3", "4"… for the libraries. Numbers have the advantage that we could agree that "1" is always the executable.

The problem of having the module name in the value is that we can't have duplicates, no? Ie, for the same journal message we can't have:

COREDUMP_PKGMETA_PACKAGE=fsverity:fsverity-utils
COREDUMP_PKGMETA_PACKAGE=libelf:libelf0
COREDUMP_PKGMETA_PACKAGE=libc6:glibc
...

Or can we?

Regarding the "main" one, I don't know if we can rely on the first module being found to be the executable, or if the ordering might be unstable

keszybz commented:
> Or can we?

Yes, we can ;)

bluca commented:
> > Or can we?
> 
> Yes, we can ;)

Yes indeed - TIL!

```
$ sudo journalctl -b -o verbose | grep PKGMETA
    COREDUMP_PKGMETA_PACKAGETYPE=fsverity:deb
    COREDUMP_PKGMETA_PACKAGE=fsverity:fsverity-utils
    COREDUMP_PKGMETA_PACKAGEVERSION=fsverity:1.3-1
    COREDUMP_PKGMETA_BUILDID=fsverity:ac89bf7175b04d7eec7f6544a923f45be111f0be
    COREDUMP_PKGMETA_PACKAGETYPE=/home/bluca/git/fsverity-utils/libfsverity.so.0:deb
    COREDUMP_PKGMETA_PACKAGE=/home/bluca/git/fsverity-utils/libfsverity.so.0:fsverity-utils
    COREDUMP_PKGMETA_PACKAGEVERSION=/home/bluca/git/fsverity-utils/libfsverity.so.0:1.3-1
    COREDUMP_PKGMETA_BUILDID=/home/bluca/git/fsverity-utils/libfsverity.so.0:fa40fdfb79aea84167c98ca8a89add9ac4f51069
```

> > I don't know if we can rely on the first module being found to be the executable, or if the ordering might be unstable
> 
> In the journal, the ordering is unstable. In the elf list? I'm not sure... Maybe we could use some heuristics? In /proc/self/maps, the executable is always first, and I think it pretty much has to, since it is loaded first by the linker, and the libraries are read based on the data therein.

If the value is module:value, then we don't need anything else right? As the "main" module name will match COREDUMP_EXE

bluca commented:
Updated with the suggested change to the journal field, and with a markdown doc describing the format

bluca commented:
Updated to only add `package` and `packageVersion` to the `COREDUMP_` list of fields, and the full json blurb as `COREDUMP_PKGMETA_JSON` for each module.
Also updated `coredumpctl info` to pring that and the build-id

keszybz commented on docs/COREDUMP_PACKAGE_METADATA.md:
```diff
@@ -0,0 +1,127 @@
+## Motivation
+
+ELF binaries get stamped with a unique, build-time generated hex string identifier called
+`build-id`, which gets embedded as an ELF note called `.note.gnu.build-id`.
```
Maybe add a link to https://fedoraproject.org/wiki/Releases/FeatureBuildId and some offical docs, if there are any?

bluca replied:
I don't know of any better link than that one

keszybz commented on docs/COREDUMP_PACKAGE_METADATA.md:
```diff
@@ -0,0 +1,127 @@
+
+ELF binaries get stamped with a unique, build-time generated hex string identifier called
+`build-id`, which gets embedded as an ELF note called `.note.gnu.build-id`.
+In most cases, this allows to associated a stripped binary with its debugging symbols.
```
associated → associate
debugging symbols → debugging information?

keszybz commented on docs/COREDUMP_PACKAGE_METADATA.md:
```diff
@@ -0,0 +1,127 @@
+In most cases, this allows to associated a stripped binary with its debugging symbols.
+It is used, for example, to dynamically fetch DWARF symbols from a debuginfo server, or
+to query the local package manager and find out the package metadata or, again, the DWARF
+symbols.
```
, or program sources.

keszybz commented on docs/COREDUMP_PACKAGE_METADATA.md:
```diff
@@ -0,0 +1,127 @@
+unable to reach a remote server.
+
+When reaching out is not possible, the packaging system local configuration state will
+hold more information - which package the binary came from, the version, and so on.
```
- → —

keszybz commented on docs/COREDUMP_PACKAGE_METADATA.md:
```diff
@@ -0,0 +1,127 @@
+manager local state configured, which is increasingly not the case, for example in case
+of minimalistic read-only images that are built and updated by a distro, but as a
+monolithic image that is updated as a single artifact, without any local packaging state
+to save space and memory.
```
Hmm, overall, this Motivation section seems a bit too blog-like. Our docs here are more of a reference documentation. I'd prefer a more "strict" description of the fact that buid-id requires either local metadata, or connection to an external servers, and that those might be either forbidden, or not available for locally installed containers.

keszybz commented on docs/COREDUMP_PACKAGE_METADATA.md:
```diff
@@ -0,0 +1,127 @@
+so on.
+
+The metadata will be embedded in a single, new ELF header section, in a key-value JSON
+format. Implementers working on parsing core files should take care in not assuming
```
should take care in not assuming … → should not assume a specific list list of keys

keszybz commented on man/systemd-coredump.xml:
```diff
@@ -352,6 +352,20 @@ flags:   ...
+
+        <listitem><para>If the executable contained .package metadata ELF notes, they will be
+        parsed and attached. The <varname>package</varname> and <varname>packageVersion</varname>
+        of the 'main' ELF module (ie: the excutable) will be appended individually, and the
+        JSON-formatted content of each module will be appended as a single entry. For more
```
Hmm, should we instead make one big json object:
```
{ 'module:fsverity': { ...},
  'module:foobar' : {... },
}
```
? This will be just less work for us, and less work for the consumer.

bluca replied:
Ok, done:

```
 COREDUMP_PKGMETA_PACKAGE=fsverity-utils
 COREDUMP_PKGMETA_PACKAGEVERSION=1.3-1
 COREDUMP_PKGMETA_JSON={"/home/bluca/git/fsverity-utils/libfsverity.so.0":{"packageType":"deb","package":"fsverity-utils","packageVersion":"1.3-1","buildid":"fa40fdfb79aea84167c98ca8a89add9ac4f51069"},"fsverity":{"packageType":"deb","package":"fsverity-utils","packageVersion":"1.3-1","buildid":"ac89bf7175b04d7eec7f6544a923f45be111f0be"}}
```

bluca commented:
Updated as requested, PTAL!

keszybz commented:
OK, let's merge this. I'm sure we'll find something or other to change here. But it's good we're early in the release cycle and have a lot of freedom to tweak things.

poettering commented on src/coredump/stacktrace.c:
```diff
@@ -113,40 +117,209 @@ static int thread_callback(Dwfl_Thread *thread, void *userdata) {
+                                        log_oom();
+                                        return DWARF_CB_ABORT;
+                                }
+                                r = set_ensure_consume(c->modules, &string_hash_ops, name_key);
```
use `set_put_strdup_full()` – which combins the strdup() and the set_ensure_consume()

poettering commented on src/coredump/stacktrace.c:
```diff
@@ -113,40 +117,209 @@ static int thread_callback(Dwfl_Thread *thread, void *userdata) {
+                 * anyway to find the package metadata. It's unlikely to have the latter
+                 * without the former, but there's no hard rule. */
+                fprintf(c->f, "Found module %s without build-id\n", name);
+        } else {
```
no {} around single-line if blocks.

maybe add full stop to end of sentence in the fprintf() message

poettering commented on src/coredump/stacktrace.c:
```diff
@@ -113,40 +117,209 @@ static int thread_callback(Dwfl_Thread *thread, void *userdata) {
+                if (r < 0) {
+                        log_error_errno(r, "json_parse on %s failed: %m", id_hex_prefixed);
+                        return DWARF_CB_ABORT;
+                }
```
hmpf. please use:

```c
r = json_build(&id_json, JSON_BUILD_OBECT(JSON_BUILD_PAIR("buildid", JSON_BUILD_HEX(id, id_len))));
```

and drop the whole block here, please.

poettering replied:
hmm, shouldn't the field be named "buildId"?

poettering commented on src/coredump/coredump.c:
```diff
@@ -781,6 +783,67 @@ static int submit_coredump(
+                        continue;
+
+                /* Cannot nest two JSON_VARIANT_OBJECT_FOREACH as they define the same
+                 * iterator variable '_state' */
```
would prefer if that was fixed, by using the UNIQ macro

poettering commented on src/coredump/coredump.c:
```diff
@@ -781,6 +783,67 @@ static int submit_coredump(
+                        key_upper = strdup(key);
+                        if (!key_upper)
+                                return log_oom();
+                        key_upper = ascii_strupper(key_upper);
```
hmm, shuldn't this be COREDUMP_PKGMETA_PACKAGE_VERSION, i.e. with an extra underscore?

poettering replied:
Hmmm, the PKGMETA thing sounds less than ideal to me. All this is metadata, and I am not a fan of unnecessary abbreviations.

Maybe:

COREDUMP_PACKAGE_VERSION= and COREDUMP_PACKAGE_NAME=?

poettering commented on src/coredump/coredump.c:
```diff
@@ -781,6 +783,67 @@ static int submit_coredump(
+                                return log_oom();
+
+                        (void) iovw_put_string_field(iovw, metadata_id, json_variant_string(w));
+                }
```
instead of iterating through this, maybe do:

```c
pn = json_variant_by_key(module_json, "package");
if (pn)
        iovw_put_string_field(iovw, "COREDUMP_PKGMETA_PACKAGE", jaon_variant_string(pn));

pv =  json_variant_by_key(module_json, "packageVersion");
if (pv)
        iovw_put_string_field(iovw, "COREDUMP_PKGMETA_PACKAGE_VERSION", jaon_variant_string(pv));
```

poettering commented on src/coredump/coredumpctl.c:
```diff
@@ -716,6 +720,68 @@ static int print_info(FILE *file, sd_journal *j, bool need_space) {
+                r = json_parse(pkgmeta_json, 0, &v, NULL, NULL);
+                if (r < 0) {
+                        log_warning_errno(r, "json_parse on %s failed, ignoring: %m", pkgmeta_json);
+                } else {
```
no {} around single-line if blocks

poettering commented on src/coredump/coredumpctl.c:
```diff
@@ -716,6 +720,68 @@ static int print_info(FILE *file, sd_journal *j, bool need_space) {
+                                     _state2.idx < json_variant_elements(_state2.variant) && \
+                                     ({ module_name = json_variant_string(json_variant_by_index(_state2.variant, _state2.idx)); \
+                                        module_json = json_variant_by_index(_state2.variant, _state2.idx + 1); \
+                                        true; });                                  \
```
please just fix the macro to use UNIQ

poettering commented on src/coredump/coredumpctl.c:
```diff
@@ -716,6 +720,68 @@ static int print_info(FILE *file, sd_journal *j, bool need_space) {
+                                }
+
+                                /* We only print the build-id for the 'main' ELF module */
+                                if (!streq(module_basename, exe_basename))
```
maybe add a helper path_equal_filename() or so, that does this in one go

poettering commented on src/coredump/coredumpctl.c:
```diff
@@ -716,6 +720,68 @@ static int print_info(FILE *file, sd_journal *j, bool need_space) {
+                                                continue;
+
+                                        if (!streq(key, "buildid"))
+                                                continue;
```
no, please don't loop here. use json_variant_by_key()

poettering commented on docs/COREDUMP_PACKAGE_METADATA.md:
````diff
@@ -0,0 +1,110 @@
+```json
+{
+     "packageType":"rpm",          # this provides a namespace for the package+package-version fields
+     "packageDistro":"fedora",
````
why abbreviate? why not just use the proper name "Distribution"?

poettering replied:
or actually, we never use "distribution" when naming stuff, but usually "os" (e.g. os-release file and so on).

maybe hence call this: packageOperatingSystem

bluca replied:
It's how they were defined on https://github.com/systemd/systemd/issues/18433 - changed as requested

poettering commented on docs/COREDUMP_PACKAGE_METADATA.md:
```diff
@@ -0,0 +1,110 @@
+     "packageDistroVersion":"33",
+     "package":"coreutils",
+     "packageVersion": "4711.0815.fc13.arm32",
+     "cpe":                        # A CPE name for the operating system, `CPE_NAME` from os-release is a good default
```
hmm, "cpe" may also be used as an identifier for packages, see the spec. if you call this property just "cpe" this might be confusing.

maybe call this "distributionCpe" or so.

poettering commented on docs/COREDUMP_PACKAGE_METADATA.md:
```diff
@@ -0,0 +1,110 @@
+     "packageDistro":"fedora",
+     "packageDistroVersion":"33",
+     "package":"coreutils",
+     "packageVersion": "4711.0815.fc13.arm32",
```
so the whole note is already called "package", I think we can drop the prefix here hence, it's all package stuff.

hence name the props "type", "os", "osVersion", "name", version", "osCpe"?

poettering reviewed (commented):
I know I am a bit late, sorry! But this still needs ome love I think

bluca commented:
Follow-ups all applied in https://github.com/systemd/systemd/pull/19231
